backend/lib/services/members.js

```diff
@@ -18,232 +18,500 @@
+    return _.startsWith(this.id, 'system:serviceaccount:') ? 'ServiceAccount' : 'User'
+  }
+
+  get size () {
+    return 1
+  }
+
+  get roles () {
+    assert.fail(new TypeError('Getter "roles" not implemented'))
+  }
+
+  set roles (roles) {
+    assert.fail(new TypeError('Setter "roles" not implemented'))
+  }
+
+  get member () {
+    return new Member(this)
+  }
 }
 
-function createServiceaccount (client, { namespace, name, createdBy }) {
-  const body = {
-    metadata: {
-      name,
-      namespace,
-      annotations: {
-        'garden.sapcloud.io/createdBy': createdBy
-      }
+class SubjectListItemUniq extends SubjectListItem {
+  constructor (subject, index) {
+    super(index)
+    this.subject = subject
+  }
+
+  get id () {
+    const { kind, name, namespace } = this.subject
+    return kind === 'ServiceAccount'
+      ? `system:serviceaccount:${namespace}:${name}`
+      : name
+  }
+
+  get roles () {
+    const { role, roles } = this.subject
+    return _
+      .chain([])
+      .concat(role, roles)
+      .compact()
+      .uniq()
+      .value()
+  }
+
+  set roles ([role, ...roles] = []) {
+    this.subject.role = role
+    if (_.isEmpty(roles)) {
+      delete this.subject.roles
+    } else {
+      this.subject.roles = _
+        .chain(roles)
+        .compact()
+        .uniq()
+        .value()
     }
   }
-  return client.core.serviceaccounts.create(namespace, body)
 }
 
-async function deleteServiceaccount (client, { namespace, name }) {
-  try {
-    return client.core.serviceaccounts.delete(namespace, name)
-  } catch (err) {
-    if (isHttpError(err, [404, 410])) {
-      return { metadata: { name, namespace } }
+class SubjectListItemGroup extends SubjectListItem {
+  constructor (items) {
+    const index = _
+      .chain(items)
+      .map('index')
+      .min()
+      .value()
+    super(index)
+    this.items = items
+  }
+
+  get id () {
+    return _
+      .chain(this.items)
+      .head()
+      .get('id')
+      .value()
+  }
+
+  get size () {
+    return this.items.length
+  }
+
+  get roles () {
+    return _
+      .chain(this.items)
+      .flatMap('roles')
+      .uniq()
+      .value()
+  }
+
+  set roles (roles) {
+    const diff = {
+      del: _.difference(this.roles, roles),
+      add: _.difference(roles, this.roles)
+    }
+    const deleteRoles = item => {
+      item.roles = _.difference(item.roles, diff.del)
+    }
+    const removeItemWithoutRoles = items => {
+      return _.compact(_.map(items, item => {
+        if (!item.roles.length) {
+          return undefined
+        }
+        return item
+      }))
     }
-    throw err
+    const addRoles = item => {
+      item.roles = _
+        .chain(item.roles)
+        .concat(diff.add)
+        .compact()
+        .uniq()
+        .value()
+    }
+    _
+      .chain(this.items)
+      .forEach(deleteRoles)
+      .head()
+      .thru(addRoles)
+      .commit()
+
+    this.items = removeItemWithoutRoles(this.items)
```

Comment:
The idea has been to represent the list of subjects of a project. If we delete an item here the link is lost.

Reply:
The issue I experienced was that I had a user with multiple entries in the member list. Then I changed roles so that one of the subjects had no role anymore. When I now saved, I ran into an issue as role is a required field. I think it is okay to remove the entr in the subject list in this case. Or am I missing something here?